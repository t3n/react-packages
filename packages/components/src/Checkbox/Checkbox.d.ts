import React from 'react';
import { ThemeFeedbackColor } from '@t3n/theme/src/theme/colors/colors';
declare type VariantType = 'light' | 'dark';
export interface CheckboxProps {
    name: string;
    value: any;
    checked: boolean;
    label?: string;
    disabled?: boolean;
    variant?: VariantType;
    feedbackColor?: ThemeFeedbackColor;
    onChange?: (e: React.ChangeEvent<HTMLInputElement>) => void;
}
declare const Checkbox: ({ checked, onChange, label, disabled, feedbackColor, name, value, variant, }: CheckboxProps) => JSX.Element;
export default Checkbox;
