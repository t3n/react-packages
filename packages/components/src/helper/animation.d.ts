import { ThemeProps } from '@t3n/theme';
export declare const backgroundAnimation: import("styled-components").FlattenInterpolation<import("styled-components").ThemedStyledProps<ThemeProps, any>>;
