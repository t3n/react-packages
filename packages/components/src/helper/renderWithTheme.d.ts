import 'jest-styled-components';
import React from 'react';
import { RenderOptions } from '@testing-library/react';
import { Theme } from '@t3n/theme';
interface OptionalTheme {
    theme?: Theme;
}
export declare const renderWithTheme: (ui: React.ReactElement, { theme, ...options }: RenderOptions & OptionalTheme) => import("@testing-library/react").RenderResult<typeof import("@testing-library/dom/types/queries"), HTMLElement>;
export {};
