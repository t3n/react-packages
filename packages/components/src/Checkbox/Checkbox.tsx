/* eslint-disable jsx-a11y/label-has-associated-control */
import React from 'react';
import styled from 'styled-components';
import { color, space, border, ColorProps } from 'styled-system';
import { MaterialCheck } from '@t3n/icons';
import { ThemeFeedbackColor } from '@t3n/theme/src/theme/colors/colors';
import { ThemeProps } from '@t3n/theme';
import { Text } from '../Text/Text';
import { Box } from '../Box/Box';

interface StyledCheckboxProps extends ColorProps {
  disabled?: boolean;
  onChange?: (e: React.ChangeEvent<HTMLInputElement>) => void;
  label?: string;
  feedbackColor?: ThemeFeedbackColor;
}

export interface CheckboxProps extends StyledCheckboxProps {
  name: string;
  checked: boolean;
  value: any;
}

const CheckboxContainer = styled(Box)<StyledCheckboxProps>`
  position: relative;
  display: inline-block;
  ${() => space({ mr: 2 })}
`;

const HiddenCheckbox = styled.input.attrs({ type: 'checkbox' })`
  position: absolute;
  width: 1rem;
  height: 1rem;
  opacity: 0;
  ${() => space({ my: 0, mr: 2, ml: 0 })}
  cursor: ${({ disabled }) => (disabled ? 'not-allowed' : 'pointer')};
`;

const StyledCheckbox = styled(Box)<StyledCheckboxProps>`
  display: inline-block;
  position: relative;
  line-height: 1;
  width: 1rem;
  height: 1rem;
  border-radius: 2px;
  pointer-events: none;
  ${({ theme }) => color({ theme, bg: 'shades.grey204' })};
  ${({ feedbackColor, theme }: StyledCheckboxProps & ThemeProps) =>
    border({
      theme,
      border: '1px solid',
      borderColor: feedbackColor
        ? theme.colors.feedback[feedbackColor]
        : theme.colors.shades.grey42
    })};
`;

const StyledMaterialCheck = styled(MaterialCheck)<
  Omit<CheckboxProps, 'name' | 'value'>
>`
  position: absolute;
  top: -1px;
  left: -1px;
  transition: all 0.1s ease-in-out;
  pointer-events: none;
  opacity: ${({ checked }) => (checked ? 1 : 0)};
  transform: scale(${({ checked }) => (checked ? 1 : 0)});
  cursor: ${({ disabled }) => (disabled ? 'not-allowed' : 'pointer')};
`;

const StyledLabel = styled.label<StyledCheckboxProps>`
  display: inline-flex;
  align-items: center;
  line-height: 1;
  pointer-events: ${({ disabled }) => (disabled ? 'none' : 'auto')};
  cursor: ${({ disabled }) => (disabled ? 'not-allowed' : 'pointer')};
  ${({ disabled, theme }) =>
    color({ theme, color: disabled ? 'shades.grey143' : 'black' })}
`;

const PlainCheckbox = ({
  checked,
  onChange,
  disabled,
  feedbackColor,
  name,
  value
}: CheckboxProps) => {
  return (
    <CheckboxContainer>
      <HiddenCheckbox
        checked={checked}
        onChange={onChange}
        disabled={disabled}
        name={name}
        value={value}
      />
      <StyledCheckbox feedbackColor={feedbackColor}>
        <StyledMaterialCheck checked={checked} />
      </StyledCheckbox>
    </CheckboxContainer>
  );
};

export const Checkbox = ({
  checked,
  onChange,
  label,
  disabled,
  feedbackColor,
  name,
  value
}: CheckboxProps) => {
  return (
    <>
      <StyledLabel>
        <PlainCheckbox
          checked={checked}
          disabled={disabled}
          onChange={onChange}
          feedbackColor={feedbackColor}
          name={name}
          value={value}
        />
        <Text small inline>
          {label}
        </Text>
      </StyledLabel>
    </>
  );
};
