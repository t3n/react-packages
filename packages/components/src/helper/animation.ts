import { css, keyframes } from 'styled-components';

import { ThemeProps } from '@t3n/theme';

const backgroundAnimationKeyframes = ({ theme }: ThemeProps) => keyframes`
  0% {
    background-color: ${theme.colors.shades.grey204}
  }

  50% {
    background-color: ${theme.colors.shades.grey244}
  }

  100% {
    background-color: ${theme.colors.shades.grey204}
  }
`;

// eslint-disable-next-line import/prefer-default-export
export const backgroundAnimation = css`
  animation: ${backgroundAnimationKeyframes} 2s linear infinite;
`;
