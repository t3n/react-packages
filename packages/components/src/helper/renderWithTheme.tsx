/* eslint-disable import/no-extraneous-dependencies */
import 'jest-styled-components';

import React from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';

import { Theme, theme as originalTheme } from '@t3n/theme';

interface OptionalTheme {
  theme?: Theme;
}

// eslint-disable-next-line import/prefer-default-export
export const renderWithTheme = (
  ui: React.ReactElement,
  { theme, ...options }: RenderOptions & OptionalTheme
) => {
  // merge over default theme

  const mergedTheme = { ...originalTheme, ...theme };

  const Wrapper: React.FC = ({ children }) => (
    <ThemeProvider theme={mergedTheme}>{children}</ThemeProvider>
  );

  return render(ui, { wrapper: Wrapper, ...options });
};
