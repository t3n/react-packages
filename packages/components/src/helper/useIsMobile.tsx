import { Theme } from '@t3n/theme';
import { useCallback, useContext, useEffect, useState } from 'react';
import { ThemeContext } from 'styled-components';

const useIsMobile = (): boolean => {
  const [isMobile, setIsMobile] = useState(false);

  const theme: Theme = useContext(ThemeContext);
  const fontSize = window
    .getComputedStyle(window.document.body)
    .getPropertyValue('font-size');
  const mobileBreakpoint =
    parseInt(theme.breakpoints[1], 10) * parseInt(fontSize, 10);

  const testIsMobile = useCallback(() => {
    if (window.matchMedia(`(max-width: ${mobileBreakpoint}px)`).matches) {
      setIsMobile(true);
    } else {
      setIsMobile(false);
    }
  }, [mobileBreakpoint]);

  useEffect(() => {
    testIsMobile();
    window.addEventListener('resize', testIsMobile);

    return () => {
      window.removeEventListener('resize', testIsMobile);
    };
  }, [testIsMobile]);

  return isMobile;
};

export default useIsMobile;
