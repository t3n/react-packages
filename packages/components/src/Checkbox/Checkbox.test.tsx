import React from 'react';
import { renderWithTheme } from '../helper/renderWithTheme';
import { Checkbox } from './Checkbox';

test('All variants Checkbox matches snapshot', () => {
  const { container } = renderWithTheme(
    <>
      <Checkbox onChange={() => {}} checked name="check" value="checkbox" />
      <Checkbox
        onChange={() => {}}
        checked
        name="check"
        variant="dark"
        value="checkbox"
      />
    </>,
    {}
  );

  expect(container).toMatchSnapshot();
});

test('Label is present on checkbox', () => {
  const { getByText } = renderWithTheme(
    <Checkbox
      label="ein Label"
      onChange={() => {}}
      checked
      name="check"
      value="checkbox"
    />,
    {}
  );

  expect(getByText(/ein label/i)).toBeTruthy();
});

test('Checkbox has feedback styles', () => {
  const { container } = renderWithTheme(
    <>
      <Checkbox
        onChange={() => {}}
        checked
        name="check"
        value="checkbox"
        feedbackColor="error"
      />
      <Checkbox
        onChange={() => {}}
        checked
        name="check"
        value="checkbox"
        feedbackColor="notice"
      />
      <Checkbox
        onChange={() => {}}
        checked
        name="check"
        value="checkbox"
        feedbackColor="warn"
      />
      <Checkbox
        onChange={() => {}}
        checked
        name="check"
        value="checkbox"
        feedbackColor="success"
      />
      <Checkbox
        onChange={() => {}}
        variant="dark"
        checked
        name="check"
        value="checkbox"
        feedbackColor="error"
      />
      <Checkbox
        onChange={() => {}}
        variant="dark"
        checked
        name="check"
        value="checkbox"
        feedbackColor="notice"
      />
      <Checkbox
        onChange={() => {}}
        variant="dark"
        checked
        name="check"
        value="checkbox"
        feedbackColor="warn"
      />
      <Checkbox
        onChange={() => {}}
        variant="dark"
        checked
        name="check"
        value="checkbox"
        feedbackColor="success"
      />
    </>,
    {}
  );

  expect(container).toMatchSnapshot();
});
