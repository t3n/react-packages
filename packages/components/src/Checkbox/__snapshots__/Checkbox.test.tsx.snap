// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox has feedback styles 1`] = `
.c2 {
  position: absolute;
  width: 100%;
  height: 100%;
  margin: 0;
  cursor: pointer;
  opacity: 0;
}

.c1 {
  display: inline-block;
  position: relative;
  width: 1rem;
  height: 1rem;
  border-radius: 2px;
  -webkit-transition: all 0.1s ease-in-out;
  transition: all 0.1s ease-in-out;
  pointer-events: none;
  margin-right: 8px;
  background-color: #fff;
  border: 1px solid;
  border-color: #f9423a;
}

.c4 {
  display: inline-block;
  position: relative;
  width: 1rem;
  height: 1rem;
  border-radius: 2px;
  -webkit-transition: all 0.1s ease-in-out;
  transition: all 0.1s ease-in-out;
  pointer-events: none;
  margin-right: 8px;
  background-color: #fff;
  border: 1px solid;
  border-color: #5bd6ff;
}

.c6 {
  display: inline-block;
  position: relative;
  width: 1rem;
  height: 1rem;
  border-radius: 2px;
  -webkit-transition: all 0.1s ease-in-out;
  transition: all 0.1s ease-in-out;
  pointer-events: none;
  margin-right: 8px;
  background-color: #fff;
  border: 1px solid;
  border-color: #ffef4f;
}

.c8 {
  display: inline-block;
  position: relative;
  width: 1rem;
  height: 1rem;
  border-radius: 2px;
  -webkit-transition: all 0.1s ease-in-out;
  transition: all 0.1s ease-in-out;
  pointer-events: none;
  margin-right: 8px;
  background-color: #fff;
  border: 1px solid;
  border-color: #34e88f;
}

.c11 {
  display: inline-block;
  position: relative;
  width: 1rem;
  height: 1rem;
  border-radius: 2px;
  -webkit-transition: all 0.1s ease-in-out;
  transition: all 0.1s ease-in-out;
  pointer-events: none;
  margin-right: 8px;
  background-color: shades.grey44;
  border: 1px solid;
  border-color: #f9423a;
}

.c12 {
  display: inline-block;
  position: relative;
  width: 1rem;
  height: 1rem;
  border-radius: 2px;
  -webkit-transition: all 0.1s ease-in-out;
  transition: all 0.1s ease-in-out;
  pointer-events: none;
  margin-right: 8px;
  background-color: shades.grey44;
  border: 1px solid;
  border-color: #5bd6ff;
}

.c13 {
  display: inline-block;
  position: relative;
  width: 1rem;
  height: 1rem;
  border-radius: 2px;
  -webkit-transition: all 0.1s ease-in-out;
  transition: all 0.1s ease-in-out;
  pointer-events: none;
  margin-right: 8px;
  background-color: shades.grey44;
  border: 1px solid;
  border-color: #ffef4f;
}

.c14 {
  display: inline-block;
  position: relative;
  width: 1rem;
  height: 1rem;
  border-radius: 2px;
  -webkit-transition: all 0.1s ease-in-out;
  transition: all 0.1s ease-in-out;
  pointer-events: none;
  margin-right: 8px;
  background-color: shades.grey44;
  border: 1px solid;
  border-color: #34e88f;
}

.c3 {
  position: absolute;
  top: -1px;
  left: -1px;
  line-height: 0;
  -webkit-transition: all 0.1s ease-in-out;
  transition: all 0.1s ease-in-out;
  pointer-events: none;
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
  opacity: 1;
}

.c3 svg {
  fill: #f9423a;
}

.c5 {
  position: absolute;
  top: -1px;
  left: -1px;
  line-height: 0;
  -webkit-transition: all 0.1s ease-in-out;
  transition: all 0.1s ease-in-out;
  pointer-events: none;
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
  opacity: 1;
}

.c5 svg {
  fill: #5bd6ff;
}

.c7 {
  position: absolute;
  top: -1px;
  left: -1px;
  line-height: 0;
  -webkit-transition: all 0.1s ease-in-out;
  transition: all 0.1s ease-in-out;
  pointer-events: none;
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
  opacity: 1;
}

.c7 svg {
  fill: #ffef4f;
}

.c9 {
  position: absolute;
  top: -1px;
  left: -1px;
  line-height: 0;
  -webkit-transition: all 0.1s ease-in-out;
  transition: all 0.1s ease-in-out;
  pointer-events: none;
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
  opacity: 1;
}

.c9 svg {
  fill: #34e88f;
}

.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  color: black;
}

.c10 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  color: white;
}

<div>
  <label
    class="c0"
  >
    <div
      class="sc-AykKD c1"
    >
      <input
        checked=""
        class="c2"
        name="check"
        type="checkbox"
        value="checkbox"
      />
      <span
        class="c3"
      >
        <svg
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
        >
          <path
            d="M9 16.17L4.83 12l-1.42 1.41L9 19 21 7l-1.41-1.41z"
          />
        </svg>
      </span>
    </div>
  </label>
  <label
    class="c0"
  >
    <div
      class="sc-AykKD c4"
    >
      <input
        checked=""
        class="c2"
        name="check"
        type="checkbox"
        value="checkbox"
      />
      <span
        class="c5"
      >
        <svg
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
        >
          <path
            d="M9 16.17L4.83 12l-1.42 1.41L9 19 21 7l-1.41-1.41z"
          />
        </svg>
      </span>
    </div>
  </label>
  <label
    class="c0"
  >
    <div
      class="sc-AykKD c6"
    >
      <input
        checked=""
        class="c2"
        name="check"
        type="checkbox"
        value="checkbox"
      />
      <span
        class="c7"
      >
        <svg
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
        >
          <path
            d="M9 16.17L4.83 12l-1.42 1.41L9 19 21 7l-1.41-1.41z"
          />
        </svg>
      </span>
    </div>
  </label>
  <label
    class="c0"
  >
    <div
      class="sc-AykKD c8"
    >
      <input
        checked=""
        class="c2"
        name="check"
        type="checkbox"
        value="checkbox"
      />
      <span
        class="c9"
      >
        <svg
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
        >
          <path
            d="M9 16.17L4.83 12l-1.42 1.41L9 19 21 7l-1.41-1.41z"
          />
        </svg>
      </span>
    </div>
  </label>
  <label
    class="c10"
  >
    <div
      class="sc-AykKD c11"
    >
      <input
        checked=""
        class="c2"
        name="check"
        type="checkbox"
        value="checkbox"
      />
      <span
        class="c3"
      >
        <svg
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
        >
          <path
            d="M9 16.17L4.83 12l-1.42 1.41L9 19 21 7l-1.41-1.41z"
          />
        </svg>
      </span>
    </div>
  </label>
  <label
    class="c10"
  >
    <div
      class="sc-AykKD c12"
    >
      <input
        checked=""
        class="c2"
        name="check"
        type="checkbox"
        value="checkbox"
      />
      <span
        class="c5"
      >
        <svg
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
        >
          <path
            d="M9 16.17L4.83 12l-1.42 1.41L9 19 21 7l-1.41-1.41z"
          />
        </svg>
      </span>
    </div>
  </label>
  <label
    class="c10"
  >
    <div
      class="sc-AykKD c13"
    >
      <input
        checked=""
        class="c2"
        name="check"
        type="checkbox"
        value="checkbox"
      />
      <span
        class="c7"
      >
        <svg
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
        >
          <path
            d="M9 16.17L4.83 12l-1.42 1.41L9 19 21 7l-1.41-1.41z"
          />
        </svg>
      </span>
    </div>
  </label>
  <label
    class="c10"
  >
    <div
      class="sc-AykKD c14"
    >
      <input
        checked=""
        class="c2"
        name="check"
        type="checkbox"
        value="checkbox"
      />
      <span
        class="c9"
      >
        <svg
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
        >
          <path
            d="M9 16.17L4.83 12l-1.42 1.41L9 19 21 7l-1.41-1.41z"
          />
        </svg>
      </span>
    </div>
  </label>
</div>
`;
